File: 프로그래머스/2/181188. 요격 시스템/요격 시스템.py
# 그리디로 풀어본다.
# 정렬을 먼저 한다.
# 앞에서 부터 범위에 포함되면 리스트 재조정
# 범위에 포함되지 않으면 삽입

def solution(targets):
    
    targets.sort(key = lambda x: (x[0], x[1]))
    
    shoot = [targets[0]]
    
    print(shoot)
    for target in targets[1:]:
        if shoot[-1][1] > target[0]:
            shoot[-1] = [target[0], min(shoot[-1][1], target[1])]
        else:
            shoot.append(target)
    
    return len(shoot)
    
    

File: 프로그래머스/2/181188. 요격 시스템/요격 시스템.java
import java.util.*;

class Solution {
    public int solution(int[][] targets) {
        Arrays.sort(targets, Comparator.comparingInt((int[] a) -> a[0]).thenComparingInt(a -> a[1]));   
        List<Shoot> answer = new ArrayList<>();
        answer.add(new Shoot(targets[0][0], targets[0][1]));
        
        for (int i = 1; i < targets.length; i++) {
            Shoot lastShoot = answer.get(answer.size() - 1);
            if (lastShoot.end > targets[i][0]) {
                lastShoot.change(targets[i][0], Math.min(lastShoot.end, targets[i][1]));
            } else {
                answer.add(new Shoot(targets[i][0], targets[i][1]));
            }
        }
        
        
        return answer.size();
    }
    
    static class Shoot {
        int start; 
        int end;
        
        public Shoot(int start, int end) {
            this.start = start;
            this.end = end;
        }
        
        public void change(int start, int end) {
            this.start = start;
            this.end = end;
        }
    }
}